pkg/handlers/generic/lifecycle/addons/helmaddon.go

```diff
@@ -88,39 +131,115 @@ func (a *helmAddonApplier) Apply(
-		},
-		Spec: caaphv1.HelmChartProxySpec{
-			RepoURL:   a.helmChart.Repository,
-			ChartName: a.helmChart.Name,
-			ClusterSelector: metav1.LabelSelector{
-				MatchLabels: map[string]string{clusterv1.ClusterNameLabel: cluster.Name},
+	if applyOpts.valueTemplater != nil {
+		values, err = applyOpts.valueTemplater(cluster, values)
+		if err != nil {
+			return fmt.Errorf("failed to template Helm values: %w", err)
+		}
+	}
+
+	targetCluster := cluster
+	if applyOpts.targetCluster != nil {
+		targetCluster = applyOpts.targetCluster
+	}
+
+	if chartProxy == nil {
+		chartProxy = &caaphv1.HelmChartProxy{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: caaphv1.GroupVersion.String(),
+				Kind:       "HelmChartProxy",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Namespace: targetCluster.Namespace,
+				Labels: map[string]string{
+					clusterv1.ClusterNameLabel: cluster.Name,
+					ClusterNamespaceLabel:      cluster.Namespace,
+					// Label values have a maximum length of 63 characters so hash the release name to
+					// ensure it fits within the limit.
+					HelmReleaseNameHashLabel: hashReleaseName(a.config.defaultHelmReleaseName),
+				},
+				GenerateName: fmt.Sprintf("%s-", a.config.defaultHelmReleaseName),
 			},
-			ReleaseNamespace: a.config.defaultHelmReleaseNamespace,
-			ReleaseName:      a.config.defaultHelmReleaseName,
-			Version:          a.helmChart.Version,
-			ValuesTemplate:   values,
+		}
+	}
+
+	chartProxy.Spec = caaphv1.HelmChartProxySpec{
+		RepoURL:   a.helmChart.Repository,
+		ChartName: a.helmChart.Name,
+		ClusterSelector: metav1.LabelSelector{
+			MatchLabels: map[string]string{clusterv1.ClusterNameLabel: targetCluster.Name},
 		},
+		ReleaseNamespace: a.config.defaultHelmReleaseNamespace,
+		ReleaseName:      a.config.defaultHelmReleaseName,
+		Version:          a.helmChart.Version,
+		ValuesTemplate:   values,
 	}
+
 	handlersutils.SetTLSConfigForHelmChartProxyIfNeeded(chartProxy)
-	if err = controllerutil.SetOwnerReference(cluster, chartProxy, a.client.Scheme()); err != nil {
+	if err = controllerutil.SetOwnerReference(targetCluster, chartProxy, a.client.Scheme()); err != nil {
 		return fmt.Errorf(
 			"failed to set owner reference on HelmChartProxy %q: %w",
 			chartProxy.Name,
 			err,
 		)
 	}
 
-	if err = k8sclient.ServerSideApply(ctx, a.client, chartProxy, k8sclient.ForceOwnership); err != nil {
-		return fmt.Errorf("failed to apply HelmChartProxy %q: %w", chartProxy.Name, err)
+	// Only server-side apply the HelmChartProxy if it already existed, i.e. that metadata.name is non-empty.
+	// This allows to use metadata.generateName for the first creation to avoid naming collisions.
+	if chartProxy.Name == "" {
+		if err = a.client.Create(ctx, chartProxy); err != nil {
+			return fmt.Errorf("failed to create HelmChartProxy %q: %w", chartProxy.Name, err)
+		}
+	} else {
+		// metadata.managedFields must be nil when using server-side apply.
+		chartProxy.ManagedFields = nil
```

Comment:
This wasn't a bug before because we weren't fetching the existing `chartProxy` first. Now that we are we need to delete the managed fields in order to use SSA here.